ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -524,24 +529,105 @@ KEXT_STATIC int HandleFileOpOperation(
                 goto CleanupAndReturn;
             }
         }
-        else
+    }
+    else if (KAUTH_FILEOP_LINK == action)
+    {
+        const char* fromPath = reinterpret_cast<const char*>(arg0);
+        const char* newPath = reinterpret_cast<const char*>(arg1);
+        
+        // TODO(Mac): We need to handle failures to lookup the vnode.  If we fail to lookup the vnode
+        // it's possible that we'll miss notifications
+        errno_t toErr = vnode_lookup(newPath, 0 /* flags */, &currentVnode, context);
+        if (0 != toErr)
         {
-            PerfSample hardLinkSample(&perfTracer, PrjFSPerfCounter_FileOp_HardLinkCreated);
+            KextLog_Error("HandleFileOpOperation: vnode_lookup failed, errno %d for path '%s'", toErr, newPath);
+            goto CleanupAndReturn;
+        }
         
-            int kauthResult;
-            int kauthError;
-            if (!ProviderMessaging_TrySendRequestAndWaitForResponse(
-                    root,
-                    MessageType_KtoU_NotifyFileHardLinkCreated,
-                    currentVnode,
-                    vnodeFsidInode,
-                    newPath,
-                    pid,
-                    procname,
-                    &kauthResult,
-                    &kauthError))
+        // Don't expect named stream here as they can't be directly hardlinked or renamed, only the main fork can
+        assert(!vnode_isnamedstream(currentVnode));
+        
+        putCurrentVnode = true;
+        
+        bool isDirectory = (0 != vnode_isdir(currentVnode));
+        
+        VirtualizationRootHandle targetRoot, fromRoot;
```

Comment:
Only run the rest of the `KAUTH_FILEOP_LINK` code if `isDirectory` is false? 

That would guarantee `MessageType_KtoU_NotifyFileHardLinkCreated` is only being delivered for files, and the `TODO: may yet be a directory` comment could be removed from PrjFSLib.cpp.

Reply:
I've also added a `KextLog_Info()` for this case so it's detectable if we go looking for issues.